src/direct-binary-upload-process.js

```diff
@@ -100,6 +100,20 @@ export default class DirectBinaryUploadProcess extends FileTransferHandler {
      * @returns {Promise} Resolves when all files have been uploaded.
      */
     async upload(uploadResult) {
+        if (typeof window !== "undefined" && typeof window.document !== "undefined") {
+            let tokenResult;
+            const origin = new URL(this.getUploadOptions().getUrl()).origin;
+            const tokenRequestUrl = `${origin}/libs/granite/csrf/token.json`;
+            const tokenRequest = new HttpRequest(this.getOptions(), tokenRequestUrl)
+                .withMethod(HttpRequest.Method.GET)
+                .withResponseType(HttpRequest.ResponseType.JSON)
+                .withUploadOptions(this.getUploadOptions());
+            const response = await this.getHttpClient().submit(tokenRequest, tokenResult);
+            this.uploadOptions.withHeaders({
+                'csrf-token': response.getData().token
+            });
+        }
```

Comment:
This will use the same CSRF token for every initiate and complete upload request. Is that correct behavior?

The implementation should be abstracted as a lower level function. Maybe add csrf.js file in https://github.com/adobe/node-httptransfer/tree/master/lib/aem to encapsulate it. The result should be the token.

Then in this function it can be added to the headers if appropriate or it would need to be requested in the initiate/complete upload functions if a unique csrf token for every request is required.

Reply:
This also needs handling for when the request fails.

Reply:
> This will use the same CSRF token for every initiate and complete upload request. Is that correct behavior?
After checkinng, it turns out that the csrf token can be reused, so discussed with Andrew and decided to put the logic 

here instead.

> The implementation should be abstracted as a lower level function. Maybe add csrf.js file in https://github.com/adobe/aem-upload/tree/master/src to encapsulate it. The result should be the token.

done

> The resulting token can be added to the headers.

done

> This also needs handling for when the request fails.

Not sure what the right behavior should be, but change it to if csrf retrieval failed throw an error and do not continue the upload. (Originally I let it continue to run in case that csrf might not be required for upload with some special condition... but maybe there isn't such case)